Support for Arbitrary Report File Header Metadata

kathlenehurt-sifive opened the pull request:
Added mechanism to specify metadata to include in report files. Also removed all implicitly added report header metadata except for the counter location.

**Example**
Report definition file:
```
header_metadata:
  period: %TS_PERIOD%
  warmup: %WARMUP_INSTS%
content:
  report:
    pattern: top 
    def_file: reports/crit_stats.yaml
    dest_file: %OUT_BASE%_time_series_crit.csv
    format: csv 
    trigger:
      start:  'core_complex.core.retire.stats.insts_retired >= %WARMUP_INSTS% -> post.WarmupTrigger'
      update-count: top.core_complex.core.retire.stats.insts_retired %TS_PERIOD%
```
Report header:
```
# counter=top.core_complex.core.retire.stats.insts_retired,period=10,warmup=29223
```

Closes #279

MichaelSchoenfelder-SiV commented:
Will this PR support this definition (header for WARMUP_INSTS, but no trigger)?
```
header_metadata:
  period: %TS_PERIOD%
  warmup: %WARMUP_INSTS%
content:
  report:
    pattern: top 
    def_file: reports/crit_stats.yaml
    dest_file: %OUT_BASE%_time_series_crit.csv
    format: csv 
    trigger:
      update-count: top.core_complex.core.retire.stats.insts_retired %TS_PERIOD%
```
It looks like it would, but that is my use case because then we could write a time series visualizer that showed the signal in both warmup and ROI, but also draw a vertical line at the beginning of the warmup period.

Another use for time series is for people to see where experiment A and B start to differ in time .  In that case, they want the time series to start at cycle zero, but would also like to know where the ROI starts.

kathlenehurt-sifive commented:
@MichaelSchoenfelder-SiV, your `update-count` example actually was not supported so I added metadata for the counters used for the start, stop and updated triggers. I'm not familiar enough with the Sparta expression triggers to know the full scope that needs to be supported so I only added metadata for counter-based triggers (only the first counter if the expression includes multiple counters).

kathlenehurt-sifive commented on sparta/src/ContextCounterTrigger.cpp:
```diff
@@ -223,17 +223,8 @@ class ContextCounterTrigger::Impl
-        if (internal_counters_.size() == 1) {
-            return internal_counters_[0];
-        }
-        throw SpartaException("You may not call the getCounter() method on this "
```
@knute-sifive are you ok with me removing this exception? None of the other triggers have exceptions like this one so I didn't think it was functionally required.

knute-sifive replied:
Did this assert fire on you?  Just curious.   Makes me scratch my head a bit -- if this assert ensures we only have 1 counter, then why is it a vector?

knute-sifive replied:
Ok, I looked at this a bit in more detail -- memory's coming back.  The context counter can be a combination of multiple counters that attribute to an expression.  When "evaluated" it can be true (fire the trigger) or false.  `getCounter` really should return the array of counters involved; and each counter's value should be considered as part of the meta data.  I _think_ 😉

kathlenehurt-sifive replied:
That's a good idea. I'll just change it to `getCounters()` that returns the whole vector. It was firing while I was trying to get the counter location to put in the header.

MichaelSchoenfelder-SiV commented:
I'd like to add 
```
header_metadata:
  period: %TS_PERIOD%
  warmup: %WARMUP_INSTS%
```
to all our existing time series reports, but we can do that in another PR.

kathlenehurt-sifive commented:
> WARMUP_INSTS

I added the header metadata section to the report for the "placeholders" test. I'll open a separate ticket to update the reports in our performance model.
